### ISSUE TITLE
Building Callgraph: RTA crashing when using dex frontend

### ISSUE THREAD

**Issue by greenerpilotx** [wala/WALA#888]
Hello,

I clone the latest Wala a few days ago to work with Android. However, I have been running into some issues using _DexFileModule_. Using the latest version of Wala, each  time I try to build a call graph using a dex file,  I get this particular exception: 

    Exception in thread "main" java.lang.ClassCastException: [Lcom.ibm.wala.dalvik.dex.instructions.Instruction; cannot be cast to [Lcom.ibm.wala.shrikeBT.IInstruction;
        at com.ibm.wala.classLoader.ShrikeIRFactory.makeIR(ShrikeIRFactory.java:51)
        at com.ibm.wala.ssa.DefaultIRFactory.makeIR(DefaultIRFactory.java:70)
        at com.ibm.wala.ssa.SSACache.findOrCreateIR(SSACache.java:69)
        at com.ibm.wala.ipa.callgraph.AnalysisCache.getIR(AnalysisCache.java:74)
        at com.ibm.wala.ipa.callgraph.propagation.cfa.ContextInsensitiveSSAInterpreter.getIR(ContextInsensitiveSSAInterpreter.java:45)
        at com.ibm.wala.ipa.callgraph.propagation.cfa.DefaultSSAInterpreter.getIR(DefaultSSAInterpreter.java:53)
        at com.ibm.wala.ipa.callgraph.propagation.cfa.DelegatingSSAContextInterpreter.getIR(DelegatingSSAContextInterpreter.java:53)
        at com.ibm.wala.ipa.callgraph.impl.ExplicitCallGraph$ExplicitNode.getIR(ExplicitCallGraph.java:295)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.getTargetPointerKey(SSAPropagationCallGraphBuilder.java:2287)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.processCallingConstraints(SSAPropagationCallGraphBuilder.java:1722)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.processResolvedCall(SSAPropagationCallGraphBuilder.java:1671)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.access$200(SSAPropagationCallGraphBuilder.java:99)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder$ConstraintVisitor.visitInvokeInternal(SSAPropagationCallGraphBuilder.java:1183)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder$ConstraintVisitor.visitInvoke(SSAPropagationCallGraphBuilder.java:1138)
        at com.ibm.wala.ssa.SSAInvokeInstruction.visit(SSAInvokeInstruction.java:94)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.addBlockInstructionConstraints(SSAPropagationCallGraphBuilder.java:272)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.addNodeInstructionConstraints(SSAPropagationCallGraphBuilder.java:249)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.unconditionallyAddConstraintsFromNode(SSAPropagationCallGraphBuilder.java:225)
        at com.ibm.wala.ipa.callgraph.propagation.SSAPropagationCallGraphBuilder.addConstraintsFromNode(SSAPropagationCallGraphBuilder.java:190)
        at com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder.addConstraintsFromNewNodes(PropagationCallGraphBuilder.java:311)
        at com.ibm.wala.ipa.callgraph.propagation.StandardSolver.solve(StandardSolver.java:56)
        at com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder.makeCallGraph(PropagationCallGraphBuilder.java:251)
        at com.uwaterloo.recommend.Main.setScope(Main.java:120)
        at com.uwaterloo.recommend.Main.main(Main.java:53)

Below, I've included the code for creating the AnalysisScope/building the call graph.  I have had no issues using this approach with jar files. However, for some reason using dex files with the latest version causes issues. 

Here, I'm setting up the AnalaysisScope with some important Android/Java jar files. 
    

    AnalysisScope scope = AnalysisScopeReader.readJavaScope("primordial.txt", new File(exclusion_path), AnalysisScopeReader.class.getClassLoader());
    scope.addInputStreamForJarToScope(ClassLoaderReference.Primordial, new FileInputStream(android_path))
    scope.setLoaderImpl(ClassLoaderReference.Application, "com.ibm.wala.dalvik.classLoader.WDexClassLoaderImpl");
    scope.addInputStreamForJarToScope(ClassLoaderReference.Application, new FileInputStream(android2_path));`
        
Here, I'm adding the target dex module to the scope and building the call graph. 

    DexFileModule dexModule = DexFileModule.make(new File(p.toString()));
    scope.addToScope(ClassLoaderReference.Application, dexModule);
    ClassHierarchy cha = ClassHierarchyFactory.make(scope);
    AnalysisOptions options = new AnalysisOptions(scope, new AllApplicationEntrypoints(scope, cha));
    CallGraphBuilder builder = Util.makeRTABuilder(options, new AnalysisCacheImpl(), cha, scope);
    CallGraph cg = builder.makeCallGraph(options, null);

I've tried this using both RTABuilder and ZeroOneCFABuilder and get the same exception each time. I'd really appreciate any suggestions for how I can make this work using dex files.

**Comment by msridhar** [comment]
The code in this class may be helpful:

https://github.com/wala/WALA/blob/67a4dfd5d40bac85fe47c20394d10a332d3a38b8/com.ibm.wala.dalvik/src/test/java/com/ibm/wala/dalvik/test/callGraph/DalvikCallGraphTestBase.java

I'm not an expert on this code but looks like you might need to set up your scope in a particular way and also the `AnalysisCache`.  See this method and what it calls:

https://github.com/wala/WALA/blob/67a4dfd5d40bac85fe47c20394d10a332d3a38b8/com.ibm.wala.dalvik/src/test/java/com/ibm/wala/dalvik/test/callGraph/DalvikCallGraphTestBase.java#L172

Does that solve the problem?  If it does, if you could paste the code you used in the end that worked, that would be great, thanks!

**Comment by linghuiluo** [comment]
I used the following code to set the analysis scope for android apk. 
```java
 scope.setLoaderImpl(ClassLoaderReference.Application, "com.ibm.wala.dalvik.classLoader.WDexClassLoaderImpl");      
// add androidJar and apkPath to scope      
scope.addToScope(ClassLoaderReference.Primordial, new JarFile(androidJar));     
scope.addToScope(ClassLoaderReference.Application, DexFileModule.make(new File(apkPath)));

```

**Comment by greenerpilotx** [comment]
Thank you both very much for your help! The key for me turned out to be this single line of code (from the DalvikCallGraphTestBase class):

     IAnalysisCacheView cache = new AnalysisCacheImpl(new DexIRFactory());

Specifically, I needed to include "new DexIRFactory()" argument. 

The following is the complete code I needed to use. This worked when using a SSAPropagationCallGraphBuilder.

    IAnalysisCacheView cache = new AnalysisCacheImpl(new DexIRFactory());
    AnalysisOptions options = new AnalysisOptions(scope, ep);
    SSAPropagationCallGraphBuilder  cgb = Util.makeVanillaZeroOneCFABuilder(Language.JAVA, options, cache, cha, scope);
    CallGraph cg = cgb.makeCallGraph(options, null); 


However, I still have issues with the RTABuilder. It works fine when I specify AllApplicationEntrypoints(). However, if I add a custom set of entrypoints, I get this exception:

    Exception in thread "main" java.lang.ClassCastException: com.ibm.wala.dalvik.classLoader.DexIMethod cannot be cast to com.ibm.wala.classLoader.ShrikeCTMethod
        at com.ibm.wala.classLoader.CodeScanner.getNewSites(CodeScanner.java:105)
        at com.ibm.wala.ipa.callgraph.propagation.rta.ContextInsensitiveRTAInterpreter.iterateNewSites(ContextInsensitiveRTAInterpreter.java:47)
        at com.ibm.wala.ipa.callgraph.propagation.cfa.DefaultSSAInterpreter.iterateNewSites(DefaultSSAInterpreter.java:68)
        at com.ibm.wala.ipa.callgraph.propagation.rta.DelegatingRTAContextInterpreter.iterateNewSites(DelegatingRTAContextInterpreter.java:52)
        at com.ibm.wala.ipa.callgraph.propagation.rta.AbstractRTABuilder.addNewConstraints(AbstractRTABuilder.java:145)
        at com.ibm.wala.ipa.callgraph.propagation.rta.AbstractRTABuilder.addConstraintsFromNode(AbstractRTABuilder.java:135)
        at com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder.addConstraintsFromNewNodes(PropagationCallGraphBuilder.java:311)
        at com.ibm.wala.ipa.callgraph.propagation.StandardSolver.solve(StandardSolver.java:56)
        at com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder.makeCallGraph(PropagationCallGraphBuilder.java:251)
        at com.uwaterloo.recommend.Main.setScope(Main.java:154)
        at com.uwaterloo.recommend.Main.main(Main.java:58)


I use the exact same code as above, except replacing the VanillaZeroOneCFABuilder with this line: 
    
    CallGraphBuilder  cgb = Util.makeRTABuilder(options, cache, cha, scope);

**Comment by msridhar** [comment]
I'm going to assume this one is addressed; please re-open if not.

**Comment by greenerpilotx** [comment]
Just wanted to note that converting the dex files to jar files and using DefaultEntrypoint rather than DexEntrypoint resolves the issue for me. (I was also getting this error when using makeZeroOneCallGraphBuilder & this happened specifically with Android 10 dex files.) The only downside to this workaround is that errors can occur during the conversion process.

**Comment by msridhar** [comment]
Sorry for not seeing there was still a bug here. It seems the RTA builder has some assumptions baked in that the incoming code is Java bytecode. If I get time I can try to take a look. In the meantime, if you want to analyze Dex directly, you can try just the 0-CFA builder for building a call graph. It's behavior should be fairly close to RTA in practice.

**Comment by msridhar** [comment]
Closed by #994
